python/mage/tgn/definitions/layers.py

```diff
@@ -0,0 +1,287 @@
+"""
+This class contains implementations of different layers
+"""
+from typing import List, Tuple, Dict
+
+import numpy as np
+import torch.nn as nn
+import torch
+
+from mage.tgn.helper.simple_mlp import MLP
+
+from mage.tgn.definitions.tgn import GraphSumDataType, GraphAttnDataType
+
+
+class TGNLayer(nn.Module):
+    """
+    Base class for all implementations
+    """
+
+    def __init__(
+        self,
+        embedding_dimension: int,
+        edge_feature_dim: int,
+        time_encoding_dim: int,
+        node_features_dim: int,
+        num_neighbors: int,
+        num_of_layers,
+    ):
+        super().__init__()
+        self.embedding_dimension = embedding_dimension
+        self.edge_feature_dim = edge_feature_dim
+        self.time_encoding_dim = time_encoding_dim
+        self.node_features_dim = node_features_dim
+        self.num_neighbors = num_neighbors
+        self.num_of_layers = num_of_layers
+
+
+class TGNLayerGraphSumEmbedding(TGNLayer):
+    """
+    TGN layer implementation of graph sum embedding
+    """
+
+    def __init__(
+        self,
+        embedding_dimension: int,
+        edge_feature_dim: int,
+        time_encoding_dim: int,
+        node_features_dim: int,
+        num_neighbors: int,
+        num_of_layers: int,
+    ):
+        super().__init__(
+            embedding_dimension,
+            edge_feature_dim,
+            time_encoding_dim,
+            node_features_dim,
+            num_neighbors,
+            num_of_layers,
+        )
+        # Initialize W1 matrix and W2 matrix
+
+        self.linear_1s = nn.ModuleList(
+            torch.nn.Linear(
+                embedding_dimension + time_encoding_dim + edge_feature_dim,
+                embedding_dimension,
+            )
+            for _ in range(self.num_of_layers)
+        )
+
+        self.linear_2s = nn.ModuleList(
+            torch.nn.Linear(
+                embedding_dimension + embedding_dimension, embedding_dimension
+            )
+            for _ in range(self.num_of_layers)
+        )
+        self.relu = torch.nn.ReLU()
+
+    def forward(self, data: GraphSumDataType):
+        (
+            node_layers,
+            mappings,
+            edge_layers,
+            neighbors_arr,
+            features,
+            edge_features,
+            time_features,
+        ) = data
+
+        out = features
+
+        for k in range(self.num_of_layers):
+            mapping = mappings[k]
+            nodes = node_layers[k + 1]  # neighbors on next layer
+            # represents how we globally gave index to node,timestamp mapping
+            global_indexes = np.array([mappings[0][(v, t)] for (v, t) in nodes])
```

Comment:
Why mappings[0] each time?

Reply:
I will resolve this conversation as we talked about this on a google meet